File: src/main/java/org/adoptopenjdk/jheappo/io/HeapProfile.java
package org.adoptopenjdk.jheappo.io;

/*
 * Copyright (c) 2018 Kirk Pepperdine.
 * Licensed under https://github.com/AdoptOpenJDK/jheappo/blob/master/LICENSE
 * Instructions: https://github.com/AdoptOpenJDK/jheappo/wiki
 */


import org.adoptopenjdk.jheappo.heapdump.*;

import java.io.*;
import java.nio.file.Path;

public class HeapProfile {


    private Path path;
    private DataInputStream input;
    private HeapProfileHeader header;
    private boolean heapDumpEnd;

    public HeapProfile(Path path) {
        this.path = path;
    }

    public void open() throws IOException {
        try {
            input = new DataInputStream(new BufferedInputStream(new FileInputStream(path.toFile())));
            heapDumpEnd = false;
        }
        catch (FileNotFoundException ex){
            System.out.println(ex.toString());
        }
    }

    public HeapProfileHeader readHeader() throws IOException {
        header = new HeapProfileHeader();
        header.extract( input);
        return header;
    }

    private byte[] readBody(DataInputStream inputStream, int bufferLength) throws IOException {
        byte[] buffer = new byte[bufferLength];
        int bytesRead = inputStream.read(buffer);
        if ( bytesRead < bufferLength) {
            heapDumpEnd = true;
            throw new IOException("bytes request exceeded bytes read");
        }
        return buffer;
    }

    public HeapProfileRecord extract() throws IOException {
        short tag = (short)input.readByte();
        long timeStamp = input.readInt();
        int bodySize = input.readInt();
        switch (tag) {
            case UTF8StringSegment.TAG :
                return new UTF8StringSegment(readBody(input, bodySize));
            case LoadClass.TAG:
                return new LoadClass(readBody(input, bodySize));
            case UnloadClass.TAG:
                System.out.println("UnloadClass");
                return new UnloadClass(readBody(input, bodySize));
            case StackFrame.TAG:
                return new StackFrame(readBody(input, bodySize));
            case StackTrace.TAG:
                return new StackTrace(readBody(input, bodySize));
            case AllocSites.TAG:
                System.out.println("AllocSites");
                return new AllocSites(readBody(input, bodySize));
            case HeapSummary.TAG:
                System.out.println("HeapSummary");
                return new HeapSummary(readBody(input, bodySize));
            case StartThread.TAG:
                System.out.println("StartThread");
                return new StartThread(readBody(input, bodySize));
            case EndThread.TAG:
                System.out.println("EndThread");
                return new EndThread(readBody(input, bodySize));
            case HeapDumpSegment.TAG1:
            case HeapDumpSegment.TAG2:
                return new HeapDumpSegment(readBody(input, bodySize));
            case HeapDumpEnd.TAG:
                System.out.println("HeapDumpEnd");
                heapDumpEnd = true;
                return new HeapDumpEnd(readBody(input, bodySize));
            case CPUSamples.TAG:
                System.out.println("CPUSamples");
                return new CPUSamples(readBody(input, bodySize));
            case ControlSettings.TAG:
                System.out.println("ControlSettings");
                return new ControlSettings(readBody(input, bodySize));
            default :
                throw new IOException("Unknown record type + " + tag);
        }
    }

    @Override
    public String toString() {
        return path.toString();
    }

    public boolean isAtHeapDumpEnd() throws IOException {
        return heapDumpEnd || input.available() == 0;
    }
}

File: src/main/java/org/adoptopenjdk/jheappo/io/HeapProfile.kt
package org.adoptopenjdk.jheappo.io

/*
 * Copyright (c) 2018 Kirk Pepperdine.
 * Licensed under https://github.com/AdoptOpenJDK/jheappo/blob/master/LICENSE
 * Instructions: https://github.com/AdoptOpenJDK/jheappo/wiki
 */


import java.io.*
import java.nio.file.Path

class HeapProfile(private val path: Path, private val input: DataInputStream) {

    companion object {
        fun open(path: Path, inputStream: InputStream): HeapProfile {
            val input = DataInputStream(BufferedInputStream(inputStream))
            return HeapProfile(path, input)
        }
    }

    private var heapDumpEnd: Boolean = false

    val isAtHeapDumpEnd: Boolean
        get() = heapDumpEnd || input.available() == 0


    fun readHeader(): HeapProfileHeader {
        val header = HeapProfileHeader()
        header.extract(input)
        return header
    }

    private fun readBody(inputStream: DataInputStream, bufferLength: Int): EncodedChunk {
        val buffer = ByteArray(bufferLength)
        // TODO allow short reads
        val bytesRead = inputStream.read(buffer)
        if (bytesRead < bufferLength) {
            heapDumpEnd = true
            throw IOException("bytes request exceeded bytes read")
        }
        return EncodedChunk(buffer)
    }

    fun extract(): HeapProfileRecord {
        val tag = EncodedChunk(byteArrayOf(input.readByte())).extractU1()
        val timeStamp = input.readInt().toLong()
        val bodySize = input.readInt()
        when (tag) {
            UTF8StringSegment.TAG -> return UTF8StringSegment(readBody(input, bodySize))
            LoadClass.TAG -> return LoadClass(readBody(input, bodySize))
            UnloadClass.TAG -> {
                println("UnloadClass")
                return UnloadClass(readBody(input, bodySize))
            }
            StackFrame.TAG -> return StackFrame(readBody(input, bodySize))
            StackTrace.TAG -> return StackTrace(readBody(input, bodySize))
            AllocSites.TAG -> {
                println("AllocSites")
                return AllocSites(readBody(input, bodySize))
            }
            HeapSummary.TAG -> {
                println("HeapSummary")
                return HeapSummary(readBody(input, bodySize))
            }
            StartThread.TAG -> {
                println("StartThread")
                return StartThread(readBody(input, bodySize))
            }
            EndThread.TAG -> {
                println("EndThread")
                return EndThread(readBody(input, bodySize))
            }
            HeapDumpSegment.TAG1, HeapDumpSegment.TAG2 -> return HeapDumpSegment(readBody(input, bodySize))
            HeapDumpEnd.TAG -> {
                println("HeapDumpEnd")
                heapDumpEnd = true
                return HeapDumpEnd(readBody(input, bodySize))
            }
            CPUSamples.TAG -> {
                println("CPUSamples")
                return CPUSamples(readBody(input, bodySize))
            }
            ControlSettings.TAG -> {
                println("ControlSettings")
                return ControlSettings(readBody(input, bodySize))
            }
            else -> throw IOException("Unknown record type + $tag")
        }
    }

    override fun toString(): String {
        return path.toString()
    }
}
